src/porepy/utils/tangential_normal_projection.py

```diff
@@ -193,74 +175,114 @@ def project_normal(self, num=None):
+            tc1 = np.zeros_like(normal)
+            negative_n1 = normal[1] < 0
+            tc1[:, negative_n1] = np.vstack(
+                [-normal[1, negative_n1], normal[0, negative_n1]]
+            )
+            positive_n1 = np.logical_not(negative_n1)
+            tc1[:, positive_n1] = np.vstack(
+                [normal[1, positive_n1], -normal[0, positive_n1]]
+            )
 
-        Returns:
-            u1': ndArray u1 / ||u1||
-            u2': ndarray (u2 - u2*u1 * u1) / ||u2||
-            u3': (optional) ndArray (u3 - u3*u2' - u3*u1')/||u3||
-        """
-        u1 = u1 / np.sqrt(np.sum(u1**2, axis=0))
-
-        u2 = u2 - np.sum(u2 * u1, axis=0) * u1
-        u2 = u2 / np.sqrt(np.sum(u2**2, axis=0))
+            basis = np.hstack([tc1, normal])
+        else:  # self.dim == 3
+            # Normalize the normal vector, just to be sure
+            normal = nc / np.linalg.norm(nc, axis=0)
+            # In 3d, there is some freedom in choosing the tangential vectors, but it is
+            # important that the definition minimize the risk for poorly conditioned
+            # computations (implementer note to self: Exactly what this means is
+            # unclear, but we have had spurious numerical issues that were traced back
+            # to a previous implementation using random vectors within the tangent
+            # plane. What caused this was never clear, but we do not want to take
+            # chances here). We choose one tangential vector to lie in the plane
+            # orthorgonal to the maximum direction of the normal vector (ex: a normal
+            # vector of [1, 3, 2] would give a first tangent vector in the xz-plane).
+            # The second tangent vector is the cross product of the normal and the first
+            # tangent vector.
+
+            # Find the maximum direction of the normal vector for each column (each
+            # individual normal vector)
+            max_dim = np.argmax(np.abs(normal), axis=0)
+            # Tangent vectors, to be filled in
+            tc1 = np.zeros_like(nc)
+            tc2 = np.zeros_like(nc)
+
+            # Loop over the dimensions, fill in all columns (individual tangent vectors)
+            # that have their maximum along this dimension.
+            for i in range(self.dim):
+                # Columns to be filled in
+                hit = max_dim == i
+                # The other dimensions
+                other_dim = np.setdiff1d(np.arange(self.dim), i)
+
+                # NOTE: We could try to assign a positive sign to the first tangent, as
+                # is done in 2d, but the benefit is less clear in 3d. Also, the special
+                # case of a normal vector aligned with a coordinate direction, where a
+                # positive direction could be most useful, is covered just below.
+                tc1[other_dim[0], hit] = -normal[other_dim[1], hit]
```

Comment:
Isn't this sort of the reverse of what we want? Assigning the two smallest values, which presumably have the largest relative round-off error? I guess I'm concerned about normal vectors almost aligned with a coordinate axis, which the special case below is the limit of.

Reply:
Good question. I think it should be safe: If the normal is almost aligned with the axis, the large number is dropped, and we have two components that are almost zero. The tangent vector will by construction be orthogonal to this 'residual' vector. 

I have added unit tests covering some cases with almost zero components, and caught no problems even for perturbations from zero that were smaller than the tolerance used in the computations.

More fundamentally, if the normal is almost aligned with an axis, we have no more information to go with than that the tangent vector should be orthogonal to this vector. You are certainly right that if the normal vector is poorly approximated, we risk incurring errors, and you may be right that small components are more likely to have rounding errors, but I don't see that we have more to work with. Am I missing something?

Reply:
One final thing, perhaps: When we normalise a small vector, my gut feeling says we may end up with something that is not exactly orthogonal anymore. Could we somehow add a check/adjustment after normalisation?

Reply:
Small numbers are not more prone to rounding errors in floating-point arithmetic than are numbers of unit size (the problems come when combining numbers of significantly different magnitudes), so although we can add a test, I do not see how this will improve matters. Let me know what you think.